### PR TITLE
[CI] Add PHP 7.2 testing container

### PR DESCRIPTION
**Pull request by LukasReschke**
Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

### CODE REVIEW

**Comment by LukasReschke** [comment]
```
Using PHP executable /usr/bin/php
PHP Fatal error:  Declaration of SebastianBergmann\Comparator\DOMNodeComparator::assertEquals($expected, $actual, $delta = 0, $canonicalize = false, $ignoreCase = false) must be compatible with SebastianBergmann\Comparator\ObjectComparator::assertEquals($expected, $actual, $delta = 0, $canonicalize = false, $ignoreCase = false, array &$processed = Array) in phar:///usr/local/bin/phpunit/sebastian-comparator/DOMNodeComparator.php on line 110
```

**Comment by LukasReschke** [comment]
With updated PHPUnit:

```
root@e7fe815f660d:/tmp/server# NOCOVERAGE=true TEST_SELECTION=NODB ./autotest.sh sqlite
Using PHP executable /usr/bin/php
Parsing all files in lib/public for the presence of @since or @deprecated on each method...

Using database oc_autotest
Setup environment for sqlite testing on local storage ...
Installing ....
Nextcloud is not installed - only a limited number of commands are available
creating sqlite db
Nextcloud was successfully installed
Testing with sqlite ...
No coverage
/usr/local/bin/phpunit --configuration phpunit-autotest.xml --exclude-group DB,SLOWDB --log-junit autotest-results-sqlite.xml
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/L10N/Factory.php on line 405
PHPUnit 5.7.21 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.0alpha3
Configuration: /tmp/server/tests/phpunit-autotest.xml

.............................................................   61 / 5389 (  1%)
.............................................................  122 / 5389 (  2%)
.............................................................  183 / 5389 (  3%)
.............................................................  244 / 5389 (  4%)
.............................................................  305 / 5389 (  5%)
.............................................................  366 / 5389 (  6%)
.............................................................  427 / 5389 (  7%)
...........................................................E.  488 / 5389 (  9%)
.............................................................  549 / 5389 ( 10%)
.............................................................  610 / 5389 ( 11%)
.............................................................  671 / 5389 ( 12%)
.............................................................  732 / 5389 ( 13%)
.............................................................  793 / 5389 ( 14%)
.........................PHP Fatal error:  Cannot use result of built-in function in write context in /tmp/server/3rdparty/pear/archive_tar/Archive/Tar.php on line 639
```

**Comment by LukasReschke** [comment]
With updated archive_tar:

```
root@e7fe815f660d:/tmp/server# NOCOVERAGE=true TEST_SELECTION=NODB ./autotest.sh sqlite
Using PHP executable /usr/bin/php
Parsing all files in lib/public for the presence of @since or @deprecated on each method...

Using database oc_autotest
Setup environment for sqlite testing on local storage ...
Installing ....
Nextcloud is not installed - only a limited number of commands are available
creating sqlite db
Nextcloud was successfully installed
Testing with sqlite ...
No coverage
/usr/local/bin/phpunit --configuration phpunit-autotest.xml --exclude-group DB,SLOWDB --log-junit autotest-results-sqlite.xml
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/Route/Route.php on line 155
PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/L10N/Factory.php on line 405
PHPUnit 5.7.21 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.0alpha3
Configuration: /tmp/server/tests/phpunit-autotest.xml

.............................................................   61 / 5389 (  1%)
.............................................................  122 / 5389 (  2%)
.............................................................  183 / 5389 (  3%)
.............................................................  244 / 5389 (  4%)
.............................................................  305 / 5389 (  5%)
.............................................................  366 / 5389 (  6%)
.............................................................  427 / 5389 (  7%)
...........................................................E.  488 / 5389 (  9%)
.............................................................  549 / 5389 ( 10%)
.............................................................  610 / 5389 ( 11%)
.............................................................  671 / 5389 ( 12%)
.............................................................  732 / 5389 ( 13%)
.............................................................  793 / 5389 ( 14%)
.............................................................  854 / 5389 ( 15%)
.............................................................  915 / 5389 ( 16%)
.............................................................  976 / 5389 ( 18%)
............................................................. 1037 / 5389 ( 19%)
............................................................. 1098 / 5389 ( 20%)
............................................................. 1159 / 5389 ( 21%)
............................................................. 1220 / 5389 ( 22%)
........................................................SSSSS 1281 / 5389 ( 23%)
SSS.......................................................... 1342 / 5389 ( 24%)
.........................................S................... 1403 / 5389 ( 26%)
....................................E........................ 1464 / 5389 ( 27%)
..EEEEE..............................EEEEEEEE........EEEEEEEE 1525 / 5389 ( 28%)
..........S...............E....EEE.EEEEEEEE.................. 1586 / 5389 ( 29%)
....................................S........................ 1647 / 5389 ( 30%)
............................................................. 1708 / 5389 ( 31%)
S............................................................ 1769 / 5389 ( 32%)
............S.....SSSSS...................................... 1830 / 5389 ( 33%)
............................................................. 1891 / 5389 ( 35%)
............................................................. 1952 / 5389 ( 36%)
............................................................. 2013 / 5389 ( 37%)
.........................EEEEEEEEE........PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/L10N/Factory.php on line 405
.PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/L10N/Factory.php on line 405
.PHP Deprecated:  Function create_function() is deprecated in /tmp/server/lib/private/L10N/Factory.php on line 405
................. 2074 / 5389 ( 38%)
............................................................. 2135 / 5389 ( 39%)
...................................PHP Fatal error:  Declaration of OC\Lockdown\Filesystem\NullStorage::copyFromStorage(OCP\Files\Storage $sourceStorage, $sourceInternalPath, $targetInternalPath) must be compatible with OC\Files\Storage\Common::copyFromStorage(OCP\Files\Storage $sourceStorage, $sourceInternalPath, $targetInternalPath, $preserveMtime = false) in /tmp/server/lib/private/Lockdown/Filesystem/NullStorage.php on line 178
```

**Comment by LukasReschke** [comment]
Lots of: 

> count(): Parameter must be an array or an object that implements Countable
> idn_to_ascii(): INTL_IDNA_VARIANT_2003 is deprecated
> Function create_function() is deprecated

**Comment by rullzer** [comment]
7.2 jobs green!!! :tada: 

https://drone.nextcloud.com/nextcloud/server/703/7/nodb-php7.2
https://drone.nextcloud.com/nextcloud/server/703/11/sqlite-php7.2

**Review by rullzer — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]

File: lib/private/Route/Route.php [hunk 1]
```diff
@@ -148,11 +148,12 @@ public function action($class, $function = null) {
...
-		$function = create_function('$param',
-			'unset($param["_route"]);'
-			.'$_GET=array_merge($_GET, $param);'
-			.'unset($param);'
-			.'require_once "'.$file.'";');
+		$function = function($param) use ($file) {
+			unset($param["_route"]);
+			$_GET=array_merge($_GET, $param);
+			unset($param);
+			require_once "'.$file.'";
```
Comment (MorrisJobke) [inline lib/private/Route/Route.php]: This is wrong 🙈 Fixed in #6057

**Pull request merged**